File: datasets/C++/Mistral/Algorithm/hard1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void matrixRotation(vector<vector<int>>& matrix, int m, int n, int r) {
    for (int layer = 0; layer < min(m, n) / 2; ++layer) {
        int firstRow = layer;
        int firstCol = layer;
        int lastRow = m - 1 - layer;
        int lastCol = n - 1 - layer;

        // Extract the elements of the current layer
        vector<int> elements;
        for (int i = firstCol; i <= lastCol; ++i) {
            elements.push_back(matrix[firstRow][i]);
        }
        for (int i = firstRow + 1; i <= lastRow; ++i) {
            elements.push_back(matrix[i][lastCol]);
        }
        for (int i = lastCol - 1; i >= firstCol; --i) {
            elements.push_back(matrix[lastRow][i]);
        }
        for (int i = lastRow - 1; i > firstRow; --i) {
            elements.push_back(matrix[i][firstCol]);
        }

        // Perform the rotation
        int length = elements.size();
        int rMod = r % length;
        rotate(elements.begin(), elements.begin() + rMod, elements.end());

        // Place the rotated elements back into the matrix
        int index = 0;
        for (int i = firstCol; i <= lastCol; ++i) {
            matrix[firstRow][i] = elements[index++];
        }
        for (int i = firstRow + 1; i <= lastRow; ++i) {
            matrix[i][lastCol] = elements[index++];
        }
        for (int i = lastCol - 1; i >= firstCol; --i) {
            matrix[lastRow][i] = elements[index++];
        }
        for (int i = lastRow - 1; i > firstRow; --i) {
            matrix[i][firstCol] = elements[index++];
        }
    }

    // Print the rotated matrix
    for (const auto& row : matrix) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }
}

int main() {
    int m, n, r;
    cin >> m >> n >> r;
    vector<vector<int>> matrix(m, vector<int>(n));
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> matrix[i][j];
        }
    }

    // Perform matrix rotation
    matrixRotation(matrix, m, n, r);

    return 0;
}

File: datasets/Python/Mistral/Algorithm/hard1.py
def matrix_rotation(matrix, m, n, r):
    for layer in range(min(m, n) // 2):
        first_row = layer
        first_col = layer
        last_row = m - 1 - layer
        last_col = n - 1 - layer

        # Extract the elements of the current layer
        elements = []
        for i in range(first_col, last_col + 1):
            elements.append(matrix[first_row][i])
        for i in range(first_row + 1, last_row + 1):
            elements.append(matrix[i][last_col])
        for i in range(last_col - 1, first_col - 1, -1):
            elements.append(matrix[last_row][i])
        for i in range(last_row - 1, first_row, -1):
            elements.append(matrix[i][first_col])

        # Perform the rotation
        length = len(elements)
        r_mod = r % length
        rotated_elements = elements[r_mod:] + elements[:r_mod]

        # Place the rotated elements back into the matrix
        index = 0
        for i in range(first_col, last_col + 1):
            matrix[first_row][i] = rotated_elements[index]
            index += 1
        for i in range(first_row + 1, last_row + 1):
            matrix[i][last_col] = rotated_elements[index]
            index += 1
        for i in range(last_col - 1, first_col - 1, -1):
            matrix[last_row][i] = rotated_elements[index]
            index += 1
        for i in range(last_row - 1, first_row, -1):
            matrix[i][first_col] = rotated_elements[index]
            index += 1

    # Print the rotated matrix
    for row in matrix:
        print(' '.join(map(str, row)))

# Read input
m, n, r = map(int, input().split())
matrix = [list(map(int, input().split())) for _ in range(m)]

# Perform matrix rotation
matrix_rotation(matrix, m, n, r)
